cpp/oneapi/dal/algo/linear_regression/backend/gpu/finalize_train_kernel_norm_eq_impl_dpc.cpp

```diff
@@ -0,0 +1,112 @@
+/*******************************************************************************
+* Copyright contributors to the oneDAL project
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/linear_regression/backend/gpu/finalize_train_kernel_norm_eq_impl.hpp"
+#include "oneapi/dal/algo/linear_regression/backend/gpu/misc.hpp"
+#include "oneapi/dal/algo/linear_regression/backend/model_impl.hpp"
+
+#include "oneapi/dal/backend/primitives/lapack.hpp"
+
+namespace oneapi::dal::linear_regression::backend {
+
+namespace be = dal::backend;
+namespace pr = be::primitives;
+
+using be::context_gpu;
+
+template <typename Float, typename Task>
+train_result<Task> finalize_train_kernel_norm_eq_impl<Float, Task>::operator()(
+    const detail::descriptor_base<Task>& desc,
+    const detail::train_parameters<Task>& params,
+    const partial_train_result<Task>& input) {
+    using dal::detail::check_mul_overflow;
+
+    using model_t = model<Task>;
+    using model_impl_t = detail::model_impl<Task>;
+
+    const bool compute_intercept = desc.get_compute_intercept();
+
+    constexpr auto uplo = pr::mkl::uplo::upper;
+    constexpr auto alloc = sycl::usm::alloc::device;
+
+    const auto response_count = input.get_partial_xty().get_row_count();
+    const auto ext_feature_count = input.get_partial_xty().get_column_count();
+    const auto feature_count = ext_feature_count - compute_intercept;
+
+    const pr::ndshape<2> xtx_shape{ ext_feature_count, ext_feature_count };
+
+    const auto xtx_nd =
+        pr::table2ndarray<Float>(q, input.get_partial_xtx(), sycl::usm::alloc::device);
+    const auto xty_nd = pr::table2ndarray<Float, pr::ndorder::f>(q,
+                                                                 input.get_partial_xty(),
+                                                                 sycl::usm::alloc::device);
+
```

Comment:
I guess you have to do:
    {
        ONEDAL_PROFILER_TASK(allreduce_xtx, q);
        comm_.allreduce(xtx_nd.flatten(q, {}), spmd::reduce_op::sum).wait();
    }
        {
        ONEDAL_PROFILER_TASK(allreduce_xty, q);
        comm_.allreduce(xty_nd.flatten(q, {}), spmd::reduce_op::sum).wait();
    }